File: compiler/server/compile.go
package server

import (
	"fmt"
	"slices"
	"strings"
	"sync"
	"wipple/codegen"
	"wipple/colors"
	"wipple/database"
	"wipple/driver"
	"wipple/feedback"
	"wipple/nodes/file"
	"wipple/syntax"
	"wipple/typecheck"
)

type CompileRequest struct {
	InputMetadata
	Code  string `json:"code,omitempty"`
	Files []File `json:"files,omitempty"`
}

type File struct {
	Path string `json:"path"`
	Code string `json:"code"`
}

type CompileResponse struct {
	Executable  string               `json:"executable,omitempty"`
	Diagnostics []ResponseDiagnostic `json:"diagnostics,omitempty"`
}

type ResponseDiagnostic struct {
	Locations []*ResponseDiagnosticLocation `json:"locations"`
	Lines     []*ResponseDiagnosticLine     `json:"lines"`
	Message   string                        `json:"message"`
}

type ResponseDiagnosticLocation struct {
	Start int `json:"start"`
	End   int `json:"end"`
	Group int `json:"group"`
	node  database.Node
}

type ResponseDiagnosticLine struct {
	Path      string                        `json:"path"`
	Source    string                        `json:"source"`
	Locations []*ResponseDiagnosticLocation `json:"locations"`
	start     int
	end       int
	offset    int
	node      database.Node
}

const defaultPath = "input"

func (request *CompileRequest) handle() (*CompileResponse, error) {
	var files []File
	if len(request.Files) > 0 {
		files = request.Files
	} else {
		files = []File{{Path: defaultPath, Code: request.Code}}
	}

	db, root, err := compile(files, request.Library)
	if err != nil {
		return nil, err
	}

	filter := func(node database.Node) bool {
		return true
	}

	seenFeedback := map[database.Node][]string{}
	feedbackItems := feedback.Collect(db, filter, func(item feedback.FeedbackItem) bool {
		if database.IsHiddenNode(item.On[0]) || !filter(item.On[0]) {
			return false
		}

		if slices.Contains(seenFeedback[item.On[0]], item.Id) {
			return false
		}

		seenFeedback[item.On[0]] = append(seenFeedback[item.On[0]], item.Id)

		return true
	})

	if len(feedbackItems) > 0 {
		responseDiagnostics := make([]ResponseDiagnostic, 0, len(feedbackItems))
		for _, item := range feedbackItems {
			slices.SortStableFunc(item.On[1:], func(left database.Node, right database.Node) int {
				return database.CompareSpans(database.GetSpanFact(left), database.GetSpanFact(right))
			})
			item.On = slices.CompactFunc(item.On, func(left database.Node, right database.Node) bool {
				return database.HaveEqualSpans(left, right)
			})

			groups := map[*typecheck.Group]int{}
			locations, lines := collectLines(db, item.On, groups)

			message := database.WrappingDisplayNode(func(node database.Node, source string) string {
				if fact, ok := database.GetFact[typecheck.TypedFact](node); ok {
					if groupIndex, ok := groups[fact.Group]; ok {
						return fmt.Sprintf("<code data-group=\"%d\">%s</code>", groupIndex, source)
					}
				}

				return "`" + source + "`"
			}, func() string {
				return item.String()
			})

			responseDiagnostics = append(responseDiagnostics, ResponseDiagnostic{
				Locations: locations,
				Lines:     lines,
				Message:   message,
			})
		}

		return &CompileResponse{Diagnostics: responseDiagnostics}, nil
	} else {
		codegen := codegen.NewCodegen(db, "index.js", codegen.Options{
			Module:    true,
			Prelude:   Prelude,
			Sourcemap: true,
			Optimize:  true,
		})

		files := make([]database.Node, 0, len(root.Files))
		for _, file := range root.Files {
			files = append(files, file)
		}

		var script string
		colors.WithoutColor(func() {
			db := db
			_ = db
			script, err = codegen.String(root, files)
		})
		if err != nil {
			return nil, err
		}

		return &CompileResponse{Executable: script}, nil
	}
}

var (
	cacheMutex = sync.RWMutex{}
	cache      = map[string]struct {
		db   *database.Db
		root *driver.RootNode
	}{}
)

func compile(files []File, libraryName string) (*database.Db, *driver.RootNode, error) {
	var db *database.Db
	var root *driver.RootNode
	if libraryName != "" {
		var err error
		db, root, err = compileLibrary(libraryName)
		if err != nil {
			return nil, nil, err
		}
	} else {
		db, root = driver.MakeRoot()
	}

	var driverFiles []*file.FileNode
	hasSyntaxError := false
	for _, f := range files {
		driverFile, err := syntax.Parse(db, f.Path, f.Code, file.ParseFile)
		if err == nil {
			driverFiles = append(driverFiles, driverFile)
		} else {
			// Syntax error is written to `db` and will be shown in diagnostics below
			hasSyntaxError = true
		}
	}

	if !hasSyntaxError {
		driver.Compile(db, root, driverFiles)
	}

	return db, root, nil
}

func compileLibrary(name string) (*database.Db, *driver.RootNode, error) {
	libraryEntry, cached, err := fetchLibrary(name)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to fetch library %q: %w", name, err)
	}

	if cached {
		cacheMutex.RLock()
		cached, ok := cache[name]
		cacheMutex.RUnlock()
		if ok {
			db := database.NewDb(cached.db)
			root := driver.CloneRoot(cached.root)
			db.Register(root)

			return db, root, nil
		}
	}

	db, root, err := compile(libraryEntry.Files, libraryEntry.Metadata.Library)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to compile library %q", name)
	}

	cacheMutex.Lock()
	cache[name] = struct {
		db   *database.Db
		root *driver.RootNode
	}{
		db:   db,
		root: root,
	}
	cacheMutex.Unlock()

	db = database.NewDb(db)
	root = driver.CloneRoot(root)
	db.Register(root)

	return db, root, nil
}

func collectLines(db *database.Db, nodes []database.Node, groups map[*typecheck.Group]int) ([]*ResponseDiagnosticLocation, []*ResponseDiagnosticLine) {
	var locations []*ResponseDiagnosticLocation
	var lines []*ResponseDiagnosticLine
	for _, node := range nodes {
		groupIndex := -1
		fact, ok := database.GetFact[typecheck.TypedFact](node)
		if ok && fact.Group != nil {
			if _, ok := groups[fact.Group]; !ok {
				groups[fact.Group] = len(groups)
			}

			groupIndex = groups[fact.Group]
		}

		span := database.GetSpanFact(node)

		if span.Path == defaultPath {
			location := ResponseDiagnosticLocation{
				Start: span.Start.Index,
				End:   span.End.Index,
				Group: groupIndex,
				node:  node,
			}

			if !slices.ContainsFunc(locations, func(other *ResponseDiagnosticLocation) bool {
				return other.Start == location.Start && other.End == location.End
			}) {
				locations = append(locations, &location)
			}

			continue
		}

		getLocation := func(offset int) *ResponseDiagnosticLocation {
			return &ResponseDiagnosticLocation{
				Start: span.Start.Column - 1 + offset,
				End:   span.End.Column + offset,
				Group: groupIndex,
				node:  node,
			}
		}

		if index := slices.IndexFunc(lines, func(line *ResponseDiagnosticLine) bool {
			return line.Path == span.Path && span.Start.Line >= line.start && span.End.Line <= line.end
		}); index != -1 {
			existing := lines[index]
			location := getLocation(existing.offset)

			if !slices.ContainsFunc(existing.Locations, func(other *ResponseDiagnosticLocation) bool {
				return other.Start == location.Start && other.End == location.End
			}) {
				existing.Locations = append(existing.Locations, location)
			}

			continue
		}

		source, ok := database.ContainsNode(db, func(node database.Node) (string, bool) {
			if _, ok := node.(*file.FileNode); !ok {
				return "", false
			}

			fileSpan := database.GetSpanFact(node)
			if fileSpan.Path == span.Path {
				return fileSpan.Source, true
			}

			return "", false
		})

		if !ok {
			continue
		}

		sourceLines := strings.Split(source, "\n")[span.Start.Line-1 : span.End.Line]
		trimmedLines := 0
		for {
			if strings.HasPrefix(strings.TrimLeft(sourceLines[0], " "), "--") {
				sourceLines = sourceLines[1:]
				trimmedLines++
			} else {
				break
			}
		}

		offset := 0
		for i := 0; i < span.End.Line-span.Start.Line-trimmedLines; i++ {
			offset += len(sourceLines[i]) + 1
		}

		lines = append(lines, &ResponseDiagnosticLine{
			Path:      span.Path,
			Source:    strings.Join(sourceLines, "\n"),
			Locations: []*ResponseDiagnosticLocation{getLocation(offset)},
			start:     span.Start.Line,
			end:       span.End.Line,
			offset:    offset,
			node:      node,
		})
	}

	locations = removeDuplicateLocations(locations)

	for _, line := range lines {
		line.Locations = removeDuplicateLocations(line.Locations)
	}

	// Delete lines that have no groups
	lines = slices.DeleteFunc(lines, func(line *ResponseDiagnosticLine) bool {
		for _, location := range line.Locations {
			if location.Group != -1 {
				return false
			}
		}

		return true
	})

	return locations, lines
}

func removeDuplicateLocations(locations []*ResponseDiagnosticLocation) []*ResponseDiagnosticLocation {
	if len(locations) == 0 {
		return nil
	}

	slices.SortStableFunc(locations, func(left *ResponseDiagnosticLocation, right *ResponseDiagnosticLocation) int {
		// Prefer non-hidden nodes
		if !database.IsHiddenNode(left.node) && database.IsHiddenNode(right.node) {
			return -1
		} else if database.IsHiddenNode(left.node) && !database.IsHiddenNode(right.node) {
			return 1
		}

		return (left.End - left.Start) - (right.End - right.Start)
	})

	locations = slices.CompactFunc(locations, func(left *ResponseDiagnosticLocation, right *ResponseDiagnosticLocation) bool {
		return left.Start == right.Start && left.End == right.End
	})

	return locations
}

File: compiler/server/compile.rs
use crate::{File, InputMetadata, compile};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::{collections::BTreeSet, sync::Arc};
use wipple::{
    codegen::{self, CodegenCtx},
    database::{Db, NodeRef, RenderConfig},
    feedback::{FeedbackItem, collect_feedback},
    nodes::FileNode,
    typecheck::{Group, Typed},
};

const INPUT_PATH: &str = "input";
const PRELUDE: &str = include_str!("../runtime/runtime.js");

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Request {
    #[serde(flatten)]
    metadata: InputMetadata,
    code: String,
}

pub async fn handle(request: Request) -> anyhow::Result<serde_json::Value> {
    let (mut db, files) = compile(
        &[File {
            path: String::from(INPUT_PATH),
            code: request.code,
        }],
        request.metadata.library.as_deref(),
    )
    .await?;

    let feedback = collect_feedback(&db, |_| true);

    if !feedback.is_empty() {
        return Ok(json!({
            "diagnostics": feedback
                .into_iter()
                .map(|item| convert_feedback(&mut db, item))
                .collect::<Vec<_>>(),
        }));
    }

    let codegen = CodegenCtx::new(
        &mut db,
        "index.js",
        codegen::Options {
            prelude: PRELUDE,
            module: true,
        },
    );

    colored::control::set_override(false);
    let script = codegen.to_string(&files)?;
    colored::control::unset_override();

    Ok(json!({ "executable": script }))
}

fn convert_feedback(db: &mut Db, item: FeedbackItem) -> serde_json::Value {
    let (groups, locations, lines) = collect_lines(db, &item.location);

    let group_count = groups.len();

    db.render_with(RenderConfig::new(move |db, value, f| {
        if let Some(node) = value.link()
            && let Some(Typed { group: Some(group) }) = db.get(node)
            && let Some(index) = groups.iter().position(|other| Arc::ptr_eq(&group, other))
        {
            write!(f, "<code data-group=\"{}\">", index)?;
            value.write(f, db)?;
            write!(f, "</code>")?;

            return Ok(());
        }

        write!(f, "`")?;
        value.write(f, db)?;
        write!(f, "`")?;

        Ok(())
    }));

    let mut message = String::new();
    (item.write)(db, &mut message);

    let graph = item
        .show_graph
        .then(|| db.filtered_graph([item.location.0.clone()]).serialize(db));

    json!({
        "locations": locations,
        "lines": lines,
        "groups": group_count,
        "message": message,
        "graph": graph,
    })
}

fn collect_lines(
    db: &Db,
    location: &(NodeRef, BTreeSet<NodeRef>),
) -> (Vec<Arc<Group>>, serde_json::Value, serde_json::Value) {
    #[derive(PartialEq, Eq, Serialize)]
    struct Location {
        start: usize,
        end: usize,
        group: i32,
    }

    #[derive(Serialize)]
    struct Line {
        source: String,
        locations: Vec<Location>,
        #[serde(skip)]
        start: usize,
        #[serde(skip)]
        end: usize,
        #[serde(skip)]
        offset: usize,
        #[serde(skip)]
        nodes: BTreeSet<NodeRef>,
    }

    let mut groups = Vec::<Arc<Group>>::new();
    let mut locations = Vec::<Location>::new();
    let mut lines = Vec::<Line>::new();
    for node in [&location.0].into_iter().chain(&location.1) {
        let mut group_index = -1;
        if let Some(Typed { group: Some(group) }) = db.get(node) {
            if let Some(index) = groups.iter().position(|other| Arc::ptr_eq(&group, other)) {
                group_index = index as i32;
            } else {
                group_index = groups.len() as i32;
                groups.push(group);
            }
        }

        let span = db.span(node);

        if span.path == INPUT_PATH {
            let location = Location {
                start: span.start.index,
                end: span.end.index,
                group: group_index,
            };

            if !locations.contains(&location) {
                locations.push(location);
            }

            continue;
        }

        let convert_offset = |offset: usize| Location {
            start: span.start.column - 1 + offset,
            end: span.end.column - 1 + offset,
            group: group_index,
        };

        if let Some(existing) = lines
            .iter_mut()
            .find(|line| span.start.line >= line.start && span.end.line <= line.end)
        {
            existing.nodes.insert(node.clone());

            let location = convert_offset(existing.offset);
            if !existing.locations.contains(&location) {
                existing.locations.push(location);
            }

            continue;
        }

        let Some(source) = db.iter_nodes().find_map(|file_node| {
            file_node.downcast_ref::<FileNode>()?;

            let file_span = db.span(&file_node);

            if file_span.path != span.path {
                return None;
            }

            Some(file_span.source)
        }) else {
            continue;
        };

        // Seek to the `span` location and trim comments
        let mut offset = 0;
        let source = source
            .lines()
            .skip(span.start.line - 1)
            .take(span.end.line - span.start.line + 1)
            .skip_while(|line| {
                if !line.trim_start().starts_with("--") {
                    return false;
                }

                offset += line.len() + 1;

                true
            })
            .collect::<Vec<_>>()
            .join("\n");

        lines.push(Line {
            source,
            locations: vec![convert_offset(offset)],
            start: span.start.line,
            end: span.end.line,
            offset,
            nodes: BTreeSet::from([node.clone()]),
        });
    }

    // Delete lines that have no groups
    lines.retain(|line| line.locations.iter().any(|location| location.group != -1));

    (
        groups,
        serde_json::to_value(locations).unwrap(),
        serde_json::to_value(lines).unwrap(),
    )
}
